### ISSUE TITLE
Server: Broken order of Netty writes

### ISSUE THREAD

**Issue by danielkec** [oracle/helidon#3640]
## Environment Details
* Helidon Version: 2.4.0
* Helidon SE
* JDK version:
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06, mixed mode, sharing)
* OS:
Distributor ID:	Ubuntu
Description:	Ubuntu 21.10
Release:	21.10
Codename:	impish
----------

## Problem Description
JsonObjects supplied with `Flow.Publisher<JsonObject>` are not always sent all. Happens intermittently.  

```java
response.send(Multi.range(0, 2)
                .flatMapCompletionStage(i -> CompletableFuture.supplyAsync(() -> JSON.createObjectBuilder().add("test", "kec").build()))
                .log(), JsonObject.class);
```

```shell
kec@romulus:~$ curl http://localhost:8080/updates
[{"test":"kec"}]
```

```
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onSubscribe(...)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(32)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onNext({"test":"kec"})
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(1)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onNext({"test":"kec"})
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇗ request(1)
2021.11.12 14:43:34 INFO Multi.log(50) !thread!:  ⇘ onComplete()
```

**Comment by danielkec** [comment]
Extra `0` byte introduced by `Charset.encode` in `io.helidon.media.jsonp.JsonpBodyWriter.JsonStructureToChunks#map` breaks Netty:
```
{"test":"kec0"} 
[123, 34, 116, 101, 115, 116, 34, 58, 34, 107, 101, 99, 48, 34, 125, 0]
java.nio.HeapByteBuffer[pos=0 lim=15 cap=16]
```
It should not be a problem because `ByteBuffer.limit=15` eg extra `0` should not be considered

But netty channel future ends exceptionally(error is swallowed):
```
io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:104)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:124)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:89)
	... 11 more 
```

**Comment by danielkec** [comment]
This is the order of writes with an extra `0` byte:
```
Write           [
Write and flush {"test":"kec0"}
Write           ,
Write and flush {"test":"kec1"}
Netty HttpObjectEncoder: DefaultHttpContent [
Netty HttpObjectEncoder: DefaultHttpContent {"test":"kec0"} 
Write           ]
Netty HttpObjectEncoder: DefaultHttpContent ]
Write last content
Netty HttpObjectEncoder: DefaultLastHttpContent 
[{"test":"kec0"}]
Netty HttpObjectEncoder: DefaultHttpContent ,
```

Versus of the order of writes when the byte is manually removed:
```
Write           [
Write           {"test":"kec0"}
Write           ,
Write           {"test":"kec1"}
Write           ]
Write last content
Netty HttpObjectEncoder: DefaultHttpContent [
Netty HttpObjectEncoder: DefaultHttpContent {"test":"kec0"}
Netty HttpObjectEncoder: DefaultHttpContent ,
Netty HttpObjectEncoder: DefaultHttpContent {"test":"kec1"}
Netty HttpObjectEncoder: DefaultHttpContent ]
Netty HttpObjectEncoder: DefaultLastHttpContent 
[{"test":"kec0"},{"test":"kec1"}]
```

Extra byte removal with: `.map(byteBuffers -> DataChunk.create(Arrays.copyOf(byteBuffers.data()[0].array(), byteBuffers.data()[0].array().length-1)))`

**Comment by danielkec** [comment]
Looks like ordering issue when mixing writes from eventloop with writes from different threads https://github.com/netty/netty/issues/3887#issuecomment-112540327

**Comment by danielkec** [comment]
Reproducer without JSON

Conditions:
- Initiate writes from different thread than eventloop
- Mixup flushes with writes:
  -  `_` - non flushed datachunks
  -  `0`,`1`,`2`,`3` - flushed datachunks

```java
        final String expected = "_1_2_3";
        DataChunk nonFlushingChunk = DataChunk.create("_".getBytes(StandardCharsets.UTF_8));

        WebServer server = WebServer.builder()
                .routing(Routing.builder()
                        .get((req, res) -> {
                            res.send(Multi.just("1", "2", "3")
                                    .map(String::valueOf)
                                    .map(String::getBytes)
                                    .map(ByteBuffer::wrap)
                                    // Condition: mixup writes with flushes
                                    .map(b -> DataChunk.create(true, b))
                                    // Condition: initiate write from different thread than event loop
                                    .observeOn(Executors.newSingleThreadExecutor())
                                    .flatMap(d -> Multi.just(nonFlushingChunk, d))
                            );
                        })
                        .build())
                .host("localhost")
                .port(0)
                .build()
                .start()
                .await();

        WebClient client = WebClient.builder()
                .baseUri("http://localhost:" + server.port())
                .build();

        for (AtomicInteger i = new AtomicInteger(); i.get() < 1000; i.getAndIncrement()) {
            client.get()
                    .request()
                    .flatMap(res -> res.content().as(String.class))
                    .forEach(s -> Assertions.assertEquals(expected, s, "Failed at " + i.get() + " " + s))
                    .await();
        }
```

**Comment by danielkec** [comment]
Issue was introduced by #3264.
Until then was reactive pipeline with `BaseResponseImpl` operating always on the same thread provided by the upstream, eg. either by Netty's event loop or other thread. By [delaying Netty's writes after request is consumed](https://github.com/oracle/helidon/commit/7440ebcf1ba7015276f383b2ee1c5904a7305711#diff-36b3438e729258fe1e48a12eacbb8f423d3a9b2632a839eeaed1ad02a778e681R224) race condition between request consumption and response reactive stream was introduced, eg. if `requestEntityAnalyzed` gets finished between `sendHeadersIfNeeded()`, `onNext` and `writeLastContent`, [Netty's write ordering guarantee](https://github.com/netty/netty/issues/3887#issuecomment-112540327) is broken.

**Comment by danielkec** [comment]
Fixed by #3679 and backported to 2x by #3671
